## Рубежный контроль №2 по курсу "Технологии машинного обучения"
### Выполнил студент 3 курса Андреев Артем, ИУ5-62
#### Вариант №1, датасет: "websites-news-technology"
#### Задание: Необходимо решить задачу классификации текстов на основе выбранного датасета. Классификация может быть бинарной или многоклассовой. Целевой признак из выбранного датасета может иметь любой физический смысл. Необходимо сформировать признаки на основе CountVectorizer или TfidfVectorizer. В качестве классификаторов необходимо использовать один из классификаторов, не относящихся к наивным Байесовским методам (например, LogisticRegression), а также Multinomial Naive Bayes (MNB), Complement Naive Bayes (CNB), Bernoulli Naive Bayes. Для каждого метода необходимо оценить качество классификации с помощью хотя бы одной метрики качества классификации (например, Accuracy). Сделате выводы о том, какой классификатор осуществляет более качественную классификацию на выбранном наборе данных.

In [83]:
import pandas as pd
import numpy as np
import warnings
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.linear_model import LogisticRegression
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.pipeline import Pipeline
from sklearn.neighbors import KNeighborsRegressor, KNeighborsClassifier
from sklearn.metrics import accuracy_score
from sklearn.naive_bayes import GaussianNB, MultinomialNB, ComplementNB, BernoulliNB

### Первичный анализ датасета

In [84]:
data = pd.read_csv('articles.csv')[:1000]
data.head()

message class
0   A former DreamWorks Animation duo lands a $30...    vr
1   The HTC Vive no longer has a waiting listHTC ...    vr
2   Indie studio hopes, but doesn’t know if Ninte...    vr
3   Now's your chance to demo the HTC Vive at sel...    vr
4   Should you buy an HTC Vive right now?The HTC ...    vr

In [85]:
data.shape

(1000, 2)

In [86]:
data.dtypes

message    object
class      object
dtype: object

In [87]:
data.isnull().sum()

message    0
class      0
dtype: int64

In [88]:
xTrain, xTest, yTrain, yTest = train_test_split(data['message'], data['class'], test_size=0.3, random_state=2)

In [89]:
def fit_and_predict_and_score(v, c):
    model = Pipeline(
        [("vectorizer", v), 
         ("classifier", c)])
    model.fit(xTrain, yTrain)
    yPredicatedTrain = model.predict(xTrain)
    yPredicatedTest = model.predict(xTest)
    print('Accuracy (train) {} %'.format(accuracy_score(yTrain, yPredicatedTrain) * 100))
    print('Accuracy (test) {} %'.format(accuracy_score(yTest, yPredicatedTest) * 100))

### Logistic Regression

In [90]:
warnings.filterwarnings('ignore')
fit_and_predict_and_score(TfidfVectorizer(), LogisticRegression(C=15))

Accuracy (train) 99.71428571428571 %
Accuracy (test) 90.33333333333333 %


### MultinomialNB

In [91]:
fit_and_predict_and_score(TfidfVectorizer(), MultinomialNB())

Accuracy (train) 79.57142857142857 %
Accuracy (test) 71.0 %


### ComplementNB

In [92]:
fit_and_predict_and_score(TfidfVectorizer(), ComplementNB())

Accuracy (train) 97.0 %
Accuracy (test) 85.0 %


### BernoulliNB

In [93]:
fit_and_predict_and_score(TfidfVectorizer(), BernoulliNB())

Accuracy (train) 88.14285714285714 %
Accuracy (test) 74.66666666666667 %


### Выводы:
* линейная регрессия работает лучше всего (90.33333333333333), 
* из наивных Байесовских методов лучше всего работает Complement Naive Bayes (85.0). 